media_driver/linux/common/ddi/media_libva_putsurface_linux.cpp

```diff
@@ -399,18 +399,19 @@ VAStatus DdiCodec_PutSurfaceLinuxHW(
     DdiMediaUtil_MediaPrintFps();
     pitch = bufferObject->iPitch;
 
-    vpCtx         = nullptr;
-    if (nullptr != mediaCtx->pVpCtxHeap->pHeapBase)
-    {
-        vpCtx = (PDDI_VP_CONTEXT)DdiMedia_GetContextFromContextID(ctx, (VAContextID)(0 + DDI_MEDIA_VACONTEXTID_OFFSET_VP), &ctxType);
-        DDI_CHK_NULL(vpCtx, "Null vpCtx", VA_STATUS_ERROR_INVALID_PARAMETER);
-        vpHal = vpCtx->pVpHal;
-        DDI_CHK_NULL(vpHal, "Null vpHal", VA_STATUS_ERROR_INVALID_PARAMETER);
-    }
-    else
+    DdiMediaUtil_LockMutex(&mediaCtx->PutSurfaceRenderMutex);
+    if (mediaCtx->PutSurfaceContext == VA_INVALID_ID)
     {
-        return VA_STATUS_ERROR_INVALID_CONTEXT;
+        VAStatus vaStatus = DdiVp_CreateContext(ctx, 0, 0, 0, 0, 0, 0, &mediaCtx->PutSurfaceContext);
+        DdiMediaUtil_UnLockMutex(&mediaCtx->PutSurfaceRenderMutex);
+        DDI_CHK_RET(vaStatus, "Create VP Context failed");
+    } else {
+        DdiMediaUtil_UnLockMutex(&mediaCtx->PutSurfaceRenderMutex);
     }
+    vpCtx = (PDDI_VP_CONTEXT)DdiMedia_GetContextFromContextID(ctx, mediaCtx->PutSurfaceContext, &ctxType);
```

Comment:
Could we move these lines into mutex protection part considering here vphal is used? Thanks.
    DdiMediaUtil_LockMutex(&mediaCtx->PutSurfaceRenderMutex);
    eStatus = vpHal->Render(&renderParams);
    if (MOS_FAILED(eStatus))
    {
        DdiMediaUtil_UnLockMutex(&mediaCtx->PutSurfaceRenderMutex);
        mos_bo_unreference(drawable_bo);
        return VA_STATUS_ERROR_OPERATION_FAILED;
    }

    DdiMediaUtil_UnLockMutex(&mediaCtx->PutSurfaceRenderMutex);

Reply:
Hi Furong,
Thanks for the review. It may not needed. 
Above lock just protect us for context creation. Once it created, it only destroy at uninitialized  time. 
The DdiMedia_GetContextFromContextID will hold a lock, and later operation like "vpHal->Render(&renderParams);" is protected by lock too.
So the entire process is thread safe.